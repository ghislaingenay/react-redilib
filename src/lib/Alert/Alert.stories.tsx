import type { StoryObj, Meta } from '@storybook/react';
import { Alert } from './Alert';
import { within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';

type TypeOfAlert = typeof Alert;
const meta: Meta<TypeOfAlert> = {
  title: 'Components/Alert',
  component: Alert,
  tags: ['autodocs'],
};

export default meta;

type Story = StoryObj<TypeOfAlert>;

export const Base: Story = {
  play: ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const PlaceholderText = canvas.getByText('Placeholder text...');
    expect(PlaceholderText).toBeInTheDocument();
  },
  argTypes: {
    type: {
      name: 'type',
      options: ['success', 'info', 'warning', 'error'],
      control: {
        type: 'select',
      },
    },
    message: {
      description: 'The message to display',
      control: {
        type: 'text',
      },
    },
    title: {
      description: 'The title to display',
      control: {
        type: 'text',
      },
    },
    haveBorder: {
      description: 'Whether to have a border or not',
      control: {
        type: 'boolean',
      },
    },
  },
};

/** Demo to show what this component can do */
export const Demo: Story = {
  name: 'Demo',
  // parameters: {
  //   controls: {
  //     exclude: /.*/,
  //   },
  // },
  decorators: [
    (Story) => (
      <div className="flex flex-col gap-10">
        <p>You select different types depending on your application needs</p>
        <Story />
      </div>
    ),
  ],
  render(args) {
    return (
      <div className="flex flex-col gap-10">
        <Alert
          type="success"
          title="Success"
          message="Please check again for more information"
        />
        <Alert
          type="info"
          title="Info"
          message="Please check again for more information"
        />
        <Alert
          type="warning"
          title="Warning"
          message="Please check again for more information"
        />
        <Alert
          type="error"
          title="Error"
          message="Please check again for more information"
        />
      </div>
    );
  },
};
