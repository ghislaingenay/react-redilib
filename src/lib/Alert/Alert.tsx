type AlertProps = {
  /** some description here */
  message: string | JSX.Element;
  title?: string;
  type: AlertType;
  border?: boolean;
};

type AlertType = 'success' | 'info' | 'warning' | 'error';

/** Some comment about the alert */
function Alert({
  message = 'Placeholder text...',
  title = 'Something',
  type = 'info',
  border,
}: AlertProps) {
  const colorBgDict: Record<
    AlertType,
    Record<'bg' | 'text' | 'bord', string>
  > = {
    success: {
      bg: 'bg-green-50',
      text: 'text-green-800',
      bord: 'border-green-800',
    },
    info: { bg: 'bg-blue-50', text: 'text-blue-800', bord: 'border-blue-800' },
    warning: {
      bg: 'bg-orange-50',
      text: 'text-orange-800',
      bord: 'border-orange-800',
    },
    error: { bg: 'bg-red-50', text: 'text-red-800', bord: 'border-red-800' },
  };

  const { bg, text, bord } = colorBgDict[type];

  const initialClass = 'flex items-center p-4 mb-4 text-sm rounded-lg mx-5';
  const borderClass = border ? `border ${bord}` : '';
  const colorClass = `${text} ${bg}`;

  const alertClass = `${initialClass} ${colorClass} ${borderClass}`;

  return (
    <div className={alertClass} role="alert">
      <svg
        className="flex-shrink-0 inline w-4 h-4 mr-3"
        aria-hidden="true"
        xmlns="http://www.w3.org/2000/svg"
        fill="currentColor"
        viewBox="0 0 20 20"
      >
        <path d="M10 .5a9.5 9.5 0 1 0 9.5 9.5A9.51 9.51 0 0 0 10 .5ZM9.5 4a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3ZM12 15H8a1 1 0 0 1 0-2h1v-3H8a1 1 0 0 1 0-2h2a1 1 0 0 1 1 1v4h1a1 1 0 0 1 0 2Z" />
      </svg>
      <span className="sr-only uppercase">{type}</span>
      <div>
        <span className="font-bold">{title}</span> {message}
      </div>
    </div>
  );
}

export { Alert };
