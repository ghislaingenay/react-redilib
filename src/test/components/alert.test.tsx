import { Alert } from '@lib/Alert';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

describe('Alert', () => {
  it('should render successfully', () => {
    const { baseElement } = render(<Alert message="Hello" type="info" />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should render successfully with title', () => {
    const { baseElement } = render(
      <Alert message="Hello" title="Title" type="info" />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
