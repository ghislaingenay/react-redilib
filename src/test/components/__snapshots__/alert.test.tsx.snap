// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert should render successfully 1`] = `
<div>
  <div
    class="flex items-center p-4 mb-4 text-sm rounded-lg mx-5 alert-info "
    role="alert"
  >
    <svg
      aria-hidden="true"
      class="flex-shrink-0 inline w-4 h-4 mr-3"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M10 .5a9.5 9.5 0 1 0 9.5 9.5A9.51 9.51 0 0 0 10 .5ZM9.5 4a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3ZM12 15H8a1 1 0 0 1 0-2h1v-3H8a1 1 0 0 1 0-2h2a1 1 0 0 1 1 1v4h1a1 1 0 0 1 0 2Z"
      />
    </svg>
    <span
      class="sr-only uppercase"
    >
      info
       alert:
    </span>
    <div
      class="font-bold tablet:font-medium"
    >
      <span
        class="font-bold hidden tablet:inline"
      >
        Something
      </span>
       
      Hello
    </div>
  </div>
</div>
`;

exports[`Alert should render successfully with title 1`] = `
<div>
  <div
    class="flex items-center p-4 mb-4 text-sm rounded-lg mx-5 alert-info "
    role="alert"
  >
    <svg
      aria-hidden="true"
      class="flex-shrink-0 inline w-4 h-4 mr-3"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M10 .5a9.5 9.5 0 1 0 9.5 9.5A9.51 9.51 0 0 0 10 .5ZM9.5 4a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3ZM12 15H8a1 1 0 0 1 0-2h1v-3H8a1 1 0 0 1 0-2h2a1 1 0 0 1 1 1v4h1a1 1 0 0 1 0 2Z"
      />
    </svg>
    <span
      class="sr-only uppercase"
    >
      info
       alert:
    </span>
    <div
      class="font-bold tablet:font-medium"
    >
      <span
        class="font-bold hidden tablet:inline"
      >
        Hello
      </span>
       
      Please try again...
    </div>
  </div>
</div>
`;
